File: 3/python3/main.py
#!/usr/bin/env python3

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # Handle the empty string case
        if not s:
            return 0

        # Initialize variables
        result = 0         # Store the length of the longest substring
        left = 0           # Left boundary of the sliding window
        current_substr = "" # To store characters in the current window

        # Traverse the string
        for right in range(len(s)):
            # If the character is already in the current substring, move the left pointer
            if s[right] in current_substr:
                # Move the left pointer to right after the first occurrence of the duplicate
                left += current_substr.index(s[right]) + 1
                # Remove all characters up to the duplicate
                current_substr = s[left:right]

            # Add the current character to the substring
            current_substr += s[right]

            # Update the result with the maximum window size found so far
            result = max(result, len(current_substr))

        return result

# Example usage:
solution = Solution()

# Test Example 1
s1 = "abcabcbb"
print("Length of longest substring in 'abcabcbb':", solution.lengthOfLongestSubstring(s1))  # Output: 3

# Test Example 2
s2 = "bbbbb"
print("Length of longest substring in 'bbbbb':", solution.lengthOfLongestSubstring(s2))  # Output: 1

# Test Example 3
s3 = "pwwkew"
print("Length of longest substring in 'pwwkew':", solution.lengthOfLongestSubstring(s3))  # Output: 3

File: 3/cpp/main.cpp
#include <iostream>
#include <string>

class Solution {
public:
    int lengthOfLongestSubstring(std::string s) {
        // Handle the empty string case
        if (s.empty()) return 0;

        // Initialize variables
        int result = 0, left = 0;
        std::string current_substr = "";

        // Traverse the string
        for (int right = 0; right < s.size(); ++right) {
            // If the character is already in the current substring, move the left pointer
            size_t pos = current_substr.find(s[right]);
            if (pos != std::string::npos) {
                // Move the left pointer to right after the first occurrence of the duplicate
                left += pos + 1;
                // Remove all characters up to the duplicate
                current_substr = current_substr.substr(pos + 1);
            }

            // Add the current character to the substring
            current_substr += s[right];

            // Update the result with the maximum window size found so far
            result = std::max(result, (int)current_substr.size());
        }

        return result;
    }
};

int main() {
    Solution sol;

    // Test case 1: "abcabcbb"
    int result1 = sol.lengthOfLongestSubstring("abcabcbb");
    std::cout << "Test 1: " << result1 << std::endl;

    // Test case 1: "bbbbb"
    int result2 = sol.lengthOfLongestSubstring("bbbbb");
    std::cout << "Test 2: " << result2 << std::endl;

    // Test case 1: "pwwkew"
    int result3 = sol.lengthOfLongestSubstring("pwwkew");
    std::cout << "Test 3: " << result3 << std::endl;

    return 0;
}
